File: FOR IOS/ECM/UI/Components/ComponentSymbol.swift
import SwiftUI

/**
 用 Canvas 画出各类元件的电路符号，不依赖任何图片资源。
 画法与安卓端 ComponentSymbol.kt 逐笔一致。
 */
struct ComponentSymbol: View {
    let type: ComponentType
    var color: Color = .white

    var body: some View {
        Canvas { context, size in
            let stroke = max(min(size.width, size.height) * 0.09, 1.2)
            drawSymbol(type, in: &context, w: size.width, h: size.height, s: stroke, c: color)
        }
        .allowsHitTesting(false)
    }
}

/// 带底色圆角方块的元件图标，用于列表行。
struct ComponentBadge: View {
    let type: ComponentType
    var size: CGFloat = 30
    var background: Color?

    var body: some View {
        RoundedRectangle(cornerRadius: size / 3.6, style: .continuous)
            .fill(background ?? type.tint)
            .frame(width: size, height: size)
            .overlay(
                ComponentSymbol(type: type, color: .white)
                    .padding(size * 0.2)
            )
    }
}

// MARK: - 绘制

private func strokeStyle(_ s: CGFloat, round: Bool = false) -> StrokeStyle {
    round ? StrokeStyle(lineWidth: s, lineCap: .round, lineJoin: .round) : StrokeStyle(lineWidth: s)
}

/// 参数化生成圆弧折线，避免各家 addArc 的方向差异。
private func arcPath(center: CGPoint, radius: CGFloat, fromDegrees: CGFloat, toDegrees: CGFloat) -> Path {
    var path = Path()
    let steps = 24
    for i in 0...steps {
        let t = fromDegrees + (toDegrees - fromDegrees) * CGFloat(i) / CGFloat(steps)
        let rad = t * .pi / 180
        let p = CGPoint(x: center.x + cos(rad) * radius, y: center.y + sin(rad) * radius)
        if i == 0 { path.move(to: p) } else { path.addLine(to: p) }
    }
    return path
}

private func line(_ a: CGPoint, _ b: CGPoint) -> Path {
    var p = Path()
    p.move(to: a)
    p.addLine(to: b)
    return p
}

private func drawSymbol(
    _ type: ComponentType,
    in context: inout GraphicsContext,
    w: CGFloat, h: CGFloat, s: CGFloat, c: Color
) {
    let midY = h / 2
    let shading = GraphicsContext.Shading.color(c)

    func stroke(_ path: Path, width: CGFloat = s, round: Bool = false) {
        context.stroke(path, with: shading, style: strokeStyle(width, round: round))
    }
    func lead(_ fromX: CGFloat, _ toX: CGFloat, y: CGFloat? = nil) {
        let yy = y ?? midY
        context.stroke(
            line(CGPoint(x: fromX, y: yy), CGPoint(x: toX, y: yy)),
            with: shading,
            style: StrokeStyle(lineWidth: s, lineCap: .round)
        )
    }
    func dot(_ center: CGPoint, _ radius: CGFloat) {
        context.fill(Path(ellipseIn: CGRect(x: center.x - radius, y: center.y - radius,
                                            width: radius * 2, height: radius * 2)), with: shading)
    }

    switch type {
    case .resistor:
        // 折线电阻符号
        var path = Path()
        let startX = w * 0.18
        let endX = w * 0.82
        let span = endX - startX
        let amp = h * 0.24
        path.move(to: CGPoint(x: 0, y: midY))
        path.addLine(to: CGPoint(x: startX, y: midY))
        let peaks = 6
        for i in 0..<peaks {
            let x = startX + span * (CGFloat(i) + 0.5) / CGFloat(peaks)
            path.addLine(to: CGPoint(x: x, y: midY + (i % 2 == 0 ? -amp : amp)))
        }
        path.addLine(to: CGPoint(x: endX, y: midY))
        path.addLine(to: CGPoint(x: w, y: midY))
        stroke(path, round: true)

    case .capacitor:
        lead(0, w * 0.42)
        lead(w * 0.58, w)
        stroke(line(CGPoint(x: w * 0.42, y: h * 0.15), CGPoint(x: w * 0.42, y: h * 0.85)))
        stroke(line(CGPoint(x: w * 0.58, y: h * 0.15), CGPoint(x: w * 0.58, y: h * 0.85)))

    case .inductor:
        lead(0, w * 0.15)
        lead(w * 0.85, w)
        let bumps = 4
        let span = w * 0.7
        let d = span / CGFloat(bumps)
        for i in 0..<bumps {
            let center = CGPoint(x: w * 0.15 + CGFloat(i) * d + d / 2, y: midY)
            stroke(arcPath(center: center, radius: d / 2, fromDegrees: 180, toDegrees: 360))
        }

    case .diode, .led:
        lead(0, w * 0.3)
        lead(w * 0.7, w)
        var tri = Path()
        tri.move(to: CGPoint(x: w * 0.3, y: h * 0.2))
        tri.addLine(to: CGPoint(x: w * 0.3, y: h * 0.8))
        tri.addLine(to: CGPoint(x: w * 0.7, y: midY))
        tri.closeSubpath()
        context.fill(tri, with: shading)
        stroke(line(CGPoint(x: w * 0.7, y: h * 0.18), CGPoint(x: w * 0.7, y: h * 0.82)))
        if type == .led {
            // 两根发光箭头
            for i in 0..<2 {
                let ox = w * (0.42 + CGFloat(i) * 0.16)
                let oy = h * 0.28
                let tip = CGPoint(x: ox + w * 0.16, y: oy - h * 0.16)
                stroke(line(CGPoint(x: ox, y: oy), tip), width: s * 0.8)
                stroke(line(tip, CGPoint(x: ox + w * 0.09, y: oy - h * 0.13)), width: s * 0.8)
                stroke(line(tip, CGPoint(x: ox + w * 0.13, y: oy - h * 0.06)), width: s * 0.8)
            }
        }

    case .transistor:
        // NPN 三极管
        stroke(line(CGPoint(x: 0, y: midY), CGPoint(x: w * 0.38, y: midY)))
        stroke(line(CGPoint(x: w * 0.38, y: h * 0.18), CGPoint(x: w * 0.38, y: h * 0.82)))
        stroke(line(CGPoint(x: w * 0.38, y: h * 0.62), CGPoint(x: w * 0.78, y: h * 0.9)))
        stroke(line(CGPoint(x: w * 0.38, y: h * 0.38), CGPoint(x: w * 0.78, y: h * 0.1)))
        stroke(line(CGPoint(x: w * 0.78, y: h * 0.9), CGPoint(x: w * 0.78, y: h)))
        stroke(line(CGPoint(x: w * 0.78, y: h * 0.1), CGPoint(x: w * 0.78, y: 0)))
        // 发射极箭头
        stroke(line(CGPoint(x: w * 0.7, y: h * 0.84), CGPoint(x: w * 0.78, y: h * 0.9)))
        stroke(line(CGPoint(x: w * 0.66, y: h * 0.92), CGPoint(x: w * 0.78, y: h * 0.9)))

    case .ic, .module, .sensor:
        let left = w * 0.24
        let right = w * 0.76
        let body = Path(roundedRect: CGRect(x: left, y: h * 0.16, width: right - left, height: h * 0.68),
                        cornerSize: CGSize(width: s, height: s))
        stroke(body)
        for i in 0..<3 {
            let y = h * (0.3 + CGFloat(i) * 0.2)
            stroke(line(CGPoint(x: w * 0.06, y: y), CGPoint(x: left, y: y)))
            stroke(line(CGPoint(x: right, y: y), CGPoint(x: w * 0.94, y: y)))
        }
        if type != .ic {
            dot(CGPoint(x: left + s * 2, y: h * 0.16 + s * 2), s * 0.9)
        }

    case .crystal:
        lead(0, w * 0.3)
        lead(w * 0.7, w)
        stroke(line(CGPoint(x: w * 0.3, y: h * 0.15), CGPoint(x: w * 0.3, y: h * 0.85)))
        stroke(line(CGPoint(x: w * 0.7, y: h * 0.15), CGPoint(x: w * 0.7, y: h * 0.85)))
        stroke(Path(CGRect(x: w * 0.4, y: h * 0.22, width: w * 0.2, height: h * 0.56)))

    case .connector:
        let shell = Path(roundedRect: CGRect(x: w * 0.12, y: h * 0.26, width: w * 0.42, height: h * 0.48),
                         cornerSize: CGSize(width: s, height: s))
        stroke(shell)
        for i in 0..<3 {
            let y = h * (0.34 + CGFloat(i) * 0.16)
            stroke(line(CGPoint(x: w * 0.54, y: y), CGPoint(x: w * 0.9, y: y)))
            dot(CGPoint(x: w * 0.9, y: y), s * 0.8)
        }

    case .switchKey:
        lead(0, w * 0.28)
        lead(w * 0.78, w)
        dot(CGPoint(x: w * 0.28, y: midY), s)
        dot(CGPoint(x: w * 0.78, y: midY), s)
        stroke(line(CGPoint(x: w * 0.28, y: midY), CGPoint(x: w * 0.74, y: h * 0.24)))

    case .power:
        // 电池符号
        lead(0, w * 0.3)
        lead(w * 0.7, w)
        stroke(line(CGPoint(x: w * 0.36, y: h * 0.16), CGPoint(x: w * 0.36, y: h * 0.84)))
        stroke(line(CGPoint(x: w * 0.5, y: h * 0.32), CGPoint(x: w * 0.5, y: h * 0.68)), width: s * 1.4)
        stroke(line(CGPoint(x: w * 0.64, y: h * 0.16), CGPoint(x: w * 0.64, y: h * 0.84)))

    case .mechanical:
        let r = w * 0.24
        stroke(Path(ellipseIn: CGRect(x: w / 2 - r, y: midY - r, width: r * 2, height: r * 2)))
        for i in 0..<6 {
            let a = CGFloat(i) * 60 * .pi / 180
            dot(CGPoint(x: w / 2 + cos(a) * w * 0.36, y: midY + sin(a) * w * 0.36), s * 0.9)
        }

    case .other:
        let box = Path(roundedRect: CGRect(x: w * 0.2, y: h * 0.2, width: w * 0.6, height: h * 0.6),
                       cornerSize: CGSize(width: s * 1.5, height: s * 1.5))
        stroke(box)
        stroke(line(CGPoint(x: w * 0.36, y: h * 0.5), CGPoint(x: w * 0.64, y: h * 0.5)))
    }
}

File: FOR Android/app/src/main/java/com/ecm/inventory/ui/components/ComponentSymbol.kt
package com.ecm.inventory.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.ecm.inventory.data.ComponentType

/**
 * 用 Canvas 画出各类元件的电路符号，不依赖任何图片资源。
 */
@Composable
fun ComponentSymbol(
    type: ComponentType,
    modifier: Modifier = Modifier,
    color: Color = Color.White
) {
    Canvas(modifier) {
        val w = size.width
        val h = size.height
        val stroke = (minOf(w, h) * 0.09f).coerceAtLeast(1.2f)
        drawSymbol(type, w, h, stroke, color)
    }
}

/** 带底色圆角方块的元件图标，用于列表行。 */
@Composable
fun ComponentBadge(
    type: ComponentType,
    modifier: Modifier = Modifier,
    size: Dp = 30.dp,
    background: Color = type.tint
) {
    Box(
        modifier
            .size(size)
            .clip(RoundedCornerShape(size / 3.6f))
            .background(background),
        contentAlignment = Alignment.Center
    ) {
        ComponentSymbol(
            type = type,
            modifier = Modifier
                .size(size)
                .padding(size * 0.2f),
            color = Color.White
        )
    }
}

private fun DrawScope.drawSymbol(type: ComponentType, w: Float, h: Float, s: Float, c: Color) {
    val midY = h / 2f
    fun lead(fromX: Float, toX: Float, y: Float = midY) =
        drawLine(c, Offset(fromX, y), Offset(toX, y), strokeWidth = s, cap = androidx.compose.ui.graphics.StrokeCap.Round)

    when (type) {
        ComponentType.RESISTOR -> {
            // 折线电阻符号
            val path = Path()
            val startX = w * 0.18f
            val endX = w * 0.82f
            val span = endX - startX
            val amp = h * 0.24f
            path.moveTo(0f, midY)
            path.lineTo(startX, midY)
            val peaks = 6
            for (i in 0 until peaks) {
                val x = startX + span * (i + 0.5f) / peaks
                path.lineTo(x, midY + if (i % 2 == 0) -amp else amp)
            }
            path.lineTo(endX, midY)
            path.lineTo(w, midY)
            drawPath(path, c, style = Stroke(width = s, cap = androidx.compose.ui.graphics.StrokeCap.Round, join = androidx.compose.ui.graphics.StrokeJoin.Round))
        }

        ComponentType.CAPACITOR -> {
            lead(0f, w * 0.42f)
            lead(w * 0.58f, w)
            drawLine(c, Offset(w * 0.42f, h * 0.15f), Offset(w * 0.42f, h * 0.85f), strokeWidth = s)
            drawLine(c, Offset(w * 0.58f, h * 0.15f), Offset(w * 0.58f, h * 0.85f), strokeWidth = s)
        }

        ComponentType.INDUCTOR -> {
            lead(0f, w * 0.15f)
            lead(w * 0.85f, w)
            val bumps = 4
            val span = w * 0.7f
            val d = span / bumps
            repeat(bumps) { i ->
                drawArc(
                    color = c,
                    startAngle = 180f,
                    sweepAngle = 180f,
                    useCenter = false,
                    topLeft = Offset(w * 0.15f + i * d, midY - d / 2f),
                    size = Size(d, d),
                    style = Stroke(width = s)
                )
            }
        }

        ComponentType.DIODE, ComponentType.LED -> {
            lead(0f, w * 0.3f)
            lead(w * 0.7f, w)
            val tri = Path().apply {
                moveTo(w * 0.3f, h * 0.2f)
                lineTo(w * 0.3f, h * 0.8f)
                lineTo(w * 0.7f, midY)
                close()
            }
            drawPath(tri, c)
            drawLine(c, Offset(w * 0.7f, h * 0.18f), Offset(w * 0.7f, h * 0.82f), strokeWidth = s)
            if (type == ComponentType.LED) {
                // 两根发光箭头
                repeat(2) { i ->
                    val ox = w * (0.42f + i * 0.16f)
                    val oy = h * 0.28f
                    drawLine(c, Offset(ox, oy), Offset(ox + w * 0.16f, oy - h * 0.16f), strokeWidth = s * 0.8f)
                    drawLine(c, Offset(ox + w * 0.16f, oy - h * 0.16f), Offset(ox + w * 0.09f, oy - h * 0.13f), strokeWidth = s * 0.8f)
                    drawLine(c, Offset(ox + w * 0.16f, oy - h * 0.16f), Offset(ox + w * 0.13f, oy - h * 0.06f), strokeWidth = s * 0.8f)
                }
            }
        }

        ComponentType.TRANSISTOR -> {
            // NPN 三极管
            drawLine(c, Offset(0f, midY), Offset(w * 0.38f, midY), strokeWidth = s)
            drawLine(c, Offset(w * 0.38f, h * 0.18f), Offset(w * 0.38f, h * 0.82f), strokeWidth = s)
            drawLine(c, Offset(w * 0.38f, h * 0.62f), Offset(w * 0.78f, h * 0.9f), strokeWidth = s)
            drawLine(c, Offset(w * 0.38f, h * 0.38f), Offset(w * 0.78f, h * 0.1f), strokeWidth = s)
            drawLine(c, Offset(w * 0.78f, h * 0.9f), Offset(w * 0.78f, h), strokeWidth = s)
            drawLine(c, Offset(w * 0.78f, h * 0.1f), Offset(w * 0.78f, 0f), strokeWidth = s)
            // 发射极箭头
            drawLine(c, Offset(w * 0.7f, h * 0.84f), Offset(w * 0.78f, h * 0.9f), strokeWidth = s)
            drawLine(c, Offset(w * 0.66f, h * 0.92f), Offset(w * 0.78f, h * 0.9f), strokeWidth = s)
        }

        ComponentType.IC, ComponentType.MODULE, ComponentType.SENSOR -> {
            val left = w * 0.24f
            val right = w * 0.76f
            drawRoundRect(
                color = c,
                topLeft = Offset(left, h * 0.16f),
                size = Size(right - left, h * 0.68f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(s, s),
                style = Stroke(width = s)
            )
            repeat(3) { i ->
                val y = h * (0.3f + i * 0.2f)
                drawLine(c, Offset(w * 0.06f, y), Offset(left, y), strokeWidth = s)
                drawLine(c, Offset(right, y), Offset(w * 0.94f, y), strokeWidth = s)
            }
            if (type != ComponentType.IC) {
                drawCircle(c, radius = s * 0.9f, center = Offset(left + s * 2f, h * 0.16f + s * 2f))
            }
        }

        ComponentType.CRYSTAL -> {
            lead(0f, w * 0.3f)
            lead(w * 0.7f, w)
            drawLine(c, Offset(w * 0.3f, h * 0.15f), Offset(w * 0.3f, h * 0.85f), strokeWidth = s)
            drawLine(c, Offset(w * 0.7f, h * 0.15f), Offset(w * 0.7f, h * 0.85f), strokeWidth = s)
            drawRect(c, topLeft = Offset(w * 0.4f, h * 0.22f), size = Size(w * 0.2f, h * 0.56f), style = Stroke(width = s))
        }

        ComponentType.CONNECTOR -> {
            drawRoundRect(
                color = c,
                topLeft = Offset(w * 0.12f, h * 0.26f),
                size = Size(w * 0.42f, h * 0.48f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(s, s),
                style = Stroke(width = s)
            )
            repeat(3) { i ->
                val y = h * (0.34f + i * 0.16f)
                drawLine(c, Offset(w * 0.54f, y), Offset(w * 0.9f, y), strokeWidth = s)
                drawCircle(c, radius = s * 0.8f, center = Offset(w * 0.9f, y))
            }
        }

        ComponentType.SWITCH -> {
            lead(0f, w * 0.28f)
            lead(w * 0.78f, w)
            drawCircle(c, radius = s, center = Offset(w * 0.28f, midY))
            drawCircle(c, radius = s, center = Offset(w * 0.78f, midY))
            drawLine(c, Offset(w * 0.28f, midY), Offset(w * 0.74f, h * 0.24f), strokeWidth = s)
        }

        ComponentType.POWER -> {
            // 电池符号
            lead(0f, w * 0.3f)
            lead(w * 0.7f, w)
            drawLine(c, Offset(w * 0.36f, h * 0.16f), Offset(w * 0.36f, h * 0.84f), strokeWidth = s)
            drawLine(c, Offset(w * 0.5f, h * 0.32f), Offset(w * 0.5f, h * 0.68f), strokeWidth = s * 1.4f)
            drawLine(c, Offset(w * 0.64f, h * 0.16f), Offset(w * 0.64f, h * 0.84f), strokeWidth = s)
        }

        ComponentType.MECHANICAL -> {
            drawCircle(c, radius = w * 0.24f, center = Offset(w / 2f, midY), style = Stroke(width = s))
            repeat(6) { i ->
                val a = Math.toRadians(i * 60.0)
                val cx = w / 2f + (Math.cos(a) * w * 0.36f).toFloat()
                val cy = midY + (Math.sin(a) * w * 0.36f).toFloat()
                drawCircle(c, radius = s * 0.9f, center = Offset(cx, cy))
            }
        }

        ComponentType.OTHER -> {
            drawRoundRect(
                color = c,
                topLeft = Offset(w * 0.2f, h * 0.2f),
                size = Size(w * 0.6f, h * 0.6f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(s * 1.5f, s * 1.5f),
                style = Stroke(width = s)
            )
            drawLine(c, Offset(w * 0.36f, h * 0.5f), Offset(w * 0.64f, h * 0.5f), strokeWidth = s)
        }
    }
}
